### DISCUSSION TITLE
Changing APIServer endpoint

### DISCUSSION THREAD

**Discussion by jaskeerat789** [kubermatic/kubeone#1945] [General]
# Description
I have a k8s cluster bootstrapped with kubeone. I used a loadbalancer for the control plane node and passed the endpoint of this LB to the kubeone config (the endpoint is replaced with a random url below)
```
apiEndpoint:
  host: 'subdomain1.domain.xyz'
  port: 6443
```

Now i decide to shift my lb to a new sub-domain and re run the `kubeone apply` command with updated `apiEndpoint`
```
apiEndpoint:
  host: 'subdomain2.domain.xyz'
  port: 6443
```

But the kubeone still tries to access the older endpoint and fails

```
INFO[16:16:30 IST] Building Kubernetes clientset...             
WARN[16:16:31 IST] Task failed, error was: unable to build dynamic client: Get "https://subdomain1.domain.xyz:6443/api?timeout=32s": ssh: rejected: connect failed (Name or service not known) 
```

**Comment by kron4eg** [comment]
@jaskeerat789 unfortunately changing apiEndpoint is not supported

**Comment by jaskeerat789** [comment]
> @jaskeerat789 unfortunately changing apiEndpoint is not supported

Is there a workaround that may or may not involve kubeone?


Also what is recommended in such scenarios?

**Comment by jaskeerat789** [comment]
Me and my team managed to achieve this with a simple ansible playbook. I think this can be made part of kubeone. I am listing all the steps we do in order to achieve this.

### On master node
- We start by stopping kubelet.
- Next we delete existing apiserver certs and key 
    - /etc/kubernetes/pki/apiserver.crt
    - /etc/kubernetes/pki/apiserver.key
- Then we update the endpoint in the config file generated by kubeone, located at `~/kubeone/cfg`. we use find and replace strategy in anisble-playbook.
- We generate new certs for the updated endpoint using the following command 
```
kubeadm init phase certs apiserver --config <config file from previous step>
```
- Next we delete all existing `.conf` files found in `/etc/kubernetes/` so that our kubecofig can be updated with the new apiserver endpoint.
- To generate new updated conf file, we run the following command
```
kubeadm init phase kubeconfig all --config <config file from 3rd step>
```
- Next we restart kubelet and upload these conf files to configmap using the following command 
```
kubeadm init phase upload-config all --config <config file from 3rd step>
```
- At last we restart the kubelet and container runtime to make sure everyone is using the updated apiserver endpoint

### On worker nodes
- We start with stopping the kubelet.
- Then we update the apiserver endpoint in `/etc/kubernetes/kubelet.conf` using find and replace strategy.
- At last we restart kubelet.
